File: 2017/spec/day11/part_1.ex
defmodule Day11 do
  def direction(str) do
    case str do
      "ne" -> {+1, 0, -1}
      "se" -> {+1, -1, 0}
      "s"  -> {0, -1, +1}
      "sw" -> {-1, 0, +1}
      "nw" -> {-1, +1, 0}
      "n"  -> {0, +1, -1}
    end
  end

  def parse_string(str) do
    String.trim(str)
    |> String.split(",")
  end

  def add_coord({x,y,z}, {x2,y2,z2}), do: {x+x2, y+y2, z+z2}

  def get_coord(str) do
    parse_string(str)
    |> Enum.map(&direction/1)
    |> Enum.reduce({0,0,0}, &add_coord/2)
  end

  def distance(inp) when is_bitstring inp do
    get_coord(inp)
    |> distance
  end

  def distance(inp) when is_tuple inp do
    Tuple.to_list(inp)
    |> Enum.map(&abs/1)
    |> Enum.max
  end

  def farthest(str) do
    parse_string(str)
    |> Enum.map(&direction/1)
    |> Enum.map_reduce({0,0,0}, fn x, acc ->
      {add_coord(x,acc), add_coord(x,acc)}end)
    |> Tuple.to_list
    |> Enum.take(1)
    |> List.flatten
    |> Enum.map(&distance/1)
    |> Enum.max
  end
end

inp = File.read!("spec/day11/steps")

Day11.distance(inp)
|> IO.inspect

Day11.farthest(inp)
|> IO.inspect

File: 2017/day11/part_1.rb
module Day11
  class Part1
    def self.run(seq)
      center = [0, 0, 0]
      location = [0, 0, 0]
      input = seq.split(",")
      input.each do |char|
        case char
        when "n"
          location[0] += 1
          location[1] -= 1
          # location[2] -= 0
        when "ne"
          # location[0] += 0
          location[1] -= 1
          location[2] += 1
        when "se"
          location[0] -= 1
          # location[1] -= 0
          location[2] += 1
        when "s"
          location[0] -= 1
          location[1] += 1
          # location[2] += 0
        when "sw"
          # location[0] -= 0
          location[1] += 1
          location[2] -= 1
        when "nw"
          location[0] += 1
          # location[1] += 0
          location[2] -= 1
        end
      end
      [(center[0] - location[0]).abs, (center[1] - location[1]).abs, (center[2] - location[2]).abs].max
    end
  end
end
